src/python/zquantum/qcbm/target_thermal_states.py

```diff
@@ -0,0 +1,235 @@
+
+    Args:
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+
+    Returns:
+       Thermal target Bistring Distribution.
+    """
+    return BitstringDistribution(
+        get_thermal_target_distribution_dict(
+            n_spins, temperature, hamiltonian_parameters
+        )
+    )
+
+
+def cumulate(
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
+) -> typing.List[float]:
+    """Generates array of cumulative probabilities from target_distribution.
+    Args:
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+
+    Returns:
+       List of cumulative probabilities from the target distribution.
+    """
+    probabilities = get_thermal_target_distribution_dict(
+        n_spins, temperature, hamiltonian_parameters
+    )
+    cumulative = []
+    tot = 0.0
+    for i in range(len(probabilities.keys())):
+        tot += list(probabilities.values())[i]
+        cumulative.append(tot)
+    return cumulative
+
+
+# Sample from the inverse cumulative partition function
+def sample(
+    n_samples: int,
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
+) -> np.array:
+    """Generates samples from the original target distribution in the form of a list of ising vectors
+    Args:
+        n_samples: the number of samples from the original distribution
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+    Returns:
+       samples: array of sample n_spin ising vectors
+    """
+    cumulative = cumulate(n_spins, temperature, hamiltonian_parameters)
+    samples = np.zeros((n_samples, n_spins))
+    y = np.zeros(n_samples)
+    for i in range(n_samples):
+        y[i] = np.random.uniform(low=0, high=1)
+    for i in range(len(y)):
+        idx = bisect.bisect_right(cumulative, y[i])
+        ising_bitstring = int2ising(idx, n_spins)
+        samples[i, :] = ising_bitstring
+    return np.asarray(samples)
+
+
+
+def get_thermal_sampled_distribution(
+    n_samples: int,
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
+) -> typing.Dict:
+    """Generates thermal states sample distribution
+    Args:
+        n_samples: the number of samples from the original distribution
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+    Returns:
+       histogram_samples: keys are binary string representations and values are corresponding probabilities.
+    """
+    sample_distribution_dict = {}
+    samples = sample(n_samples, n_spins, temperature, hamiltonian_parameters)
```

Comment:
Since I had trouble figuring out what `sample` does, it's hard to figure out what is going on in the following lines. I will wait to review this function until a later point.

Reply:
Sample just generates random samples from the distribution